libsolidity/ast/Types.h

```diff
@@ -694,11 +692,30 @@ class BoolType: public Type
 	TypeResult interfaceType(bool) const override { return this; }
 };
 
+/**
+ * Base class for types which consist of other types.
+ */
+class CompositeType: public Type
+{
+protected:
+	CompositeType() = default;
+
+public:
+	/// @returns list of all types the composite type consists of, directly or indirectly,
+	/// including the type itself, without duplicates
+	std::vector<Type const*> fullDecomposition() const;
+
+protected:
+	/// @returns list of types the composite type directly consists of
+	/// (base type for array, members for structs, value type for mappings), might contain duplicates
+	virtual std::vector<Type const*> decomposition() const = 0;
```

Comment:
maybe `subTypes` would be a better name?

Reply:
The funny thing is that it was originally called `subtypes()` but I thought that it's too vague and sounds to much like a subclass so I suggested changing it :)

Reply:
`subtypes()` sounds like something that could include e.g. both key and value type in a mapping. Or types of parameters in a function type. And this function is specifically meant to return only the types that the memory space occupied by the type can be subdivided into.

My other suggestions were: `components()`, `subdivision()`, `itemize()`, `buildingBlocks()`.

Reply:
Since for mappings it's only the value type and not the key types... maybe it's rather ``rangeTypes()`` :-D? ``rangeDecomposition()`` :-D? Well, I, again, realize that I'm terrible at naming things :-).

Reply:
I started with `subtypes` and I like it as a short colloquial word. I was afraid though that it might be read as "types derived from" (similar to [subclass](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)#Subclasses_and_superclasses), `class SubClass: SuperClass`).

On the other hand, "sub" as in "syntax tree subnode" is somewhat close to what we need.